packager/mpd/base/xml/xml_node.cc

```diff
@@ -379,98 +380,83 @@ bool RepresentationXmlNode::AddAudioInfo(const AudioInfo& audio_info) {
+  }
 
-  const bool need_segment_base = !use_segment_list && (
-      media_info.has_index_range() || media_info.has_init_range() ||
-      (media_info.has_reference_time_scale() && !media_info.has_text_info()));
-  DCHECK(!need_segment_base || !use_segment_list_text ||
-             use_segment_list);
+  XmlNode child(use_segment_list || use_single_segment_url_with_media
+                    ? "SegmentList"
+                    : "SegmentBase");
 
-  if (need_segment_base || use_segment_list_text) {
-    XmlNode child(need_segment_base ? "SegmentBase" : "SegmentList");
-    if (media_info.has_index_range()) {
-      RCHECK(child.SetStringAttribute("indexRange",
-                                      RangeToString(media_info.index_range())));
-    }
+  // Forcing SegmentList for longer audio causes sidx atom to not be
+  // generated, therefore indexRange is not added to MPD if flag is set.
+  if (media_info.has_index_range() && !use_segment_list) {
+    RCHECK(child.SetStringAttribute("indexRange",
+                                    RangeToString(media_info.index_range())));
+  }
 
-    if (media_info.has_reference_time_scale()) {
-      RCHECK(child.SetIntegerAttribute("timescale",
-                                       media_info.reference_time_scale()));
-    }
+  if (media_info.has_reference_time_scale()) {
+    RCHECK(child.SetIntegerAttribute("timescale",
+                                     media_info.reference_time_scale()));
 
-    if (media_info.has_presentation_time_offset()) {
-      RCHECK(child.SetIntegerAttribute("presentationTimeOffset",
-                                       media_info.presentation_time_offset()));
+    if (use_segment_list) {
+      const uint64_t duration_seconds = static_cast<uint64_t>(
+          floor(target_segment_duration * media_info.reference_time_scale()));
+      RCHECK(child.SetIntegerAttribute("duration", duration_seconds));
     }
+  }
 
-    if (media_info.has_init_range()) {
-      XmlNode initialization("Initialization");
-      RCHECK(initialization.SetStringAttribute(
-          "range", RangeToString(media_info.init_range())));
+  if (media_info.has_presentation_time_offset()) {
+    RCHECK(child.SetIntegerAttribute("presentationTimeOffset",
+                                     media_info.presentation_time_offset()));
+  }
 
-      RCHECK(child.AddChild(std::move(initialization)));
-    }
+  if (media_info.has_init_range()) {
+    XmlNode initialization("Initialization");
+    RCHECK(initialization.SetStringAttribute(
+        "range", RangeToString(media_info.init_range())));
 
-    if (use_segment_list_text) {
-      XmlNode media_url("SegmentURL");
-      RCHECK(
-          media_url.SetStringAttribute("media", media_info.media_file_url()));
-      RCHECK(child.AddChild(std::move(media_url)));
-    }
+    RCHECK(child.AddChild(std::move(initialization)));
+  }
 
-    RCHECK(AddChild(std::move(child)));
+  if (use_single_segment_url_with_media && !use_segment_list) {
```

Comment:
I don't think the `!use_segment_list` condition should be here. For text streams with only one segment, there should only be one `SegmentURL` regardless of whether the flag is set.

Similarly, line 421 should be:

```
if (use_segment_list && !use_single_segment_url_with_media) {
```